extensions/sql-database-projects/src/models/project.ts

```diff
@@ -1442,6 +1477,81 @@ export class Project implements ISqlProject {
 		}
 	}
 
+	/**
+	 * Adds a value to project property, separated by semicolon if
+	 * one or more values are already present in project property
+	 * @param propertyName Name of property
+	 * @param valueToAdd Value to be added to project property, values containing semicolon are not supported
+	 * @param caseSensitive Determines if value to add to collection should be compared against existing values case sensitively
+	 */
+	private async addValueToCollectionProjectProperty(propertyName: string, valueToAdd: string, caseSensitive?: boolean): Promise<void> {
+		if (valueToAdd.includes(';')) {
+			return;
+		}
+
+		let propertyValueToSet = this.evaluateProjectPropertyValue(propertyName);
+		if (propertyValueToSet === undefined) {
+			propertyValueToSet = valueToAdd;
+		} else {
+			const propertyValue = caseSensitive ? propertyValueToSet : propertyValueToSet.toLocaleLowerCase();
+			const originalValueToAdd = valueToAdd;
+			valueToAdd = caseSensitive ? valueToAdd : valueToAdd.toLocaleLowerCase();
+			if ((propertyValue.length > valueToAdd.length
+				&& (propertyValue.startsWith(`${valueToAdd};`)
+					|| propertyValue.endsWith(`;${valueToAdd}`)
+					|| propertyValue.indexOf(`;${valueToAdd};`) >= 0))
+				|| (propertyValue.length === valueToAdd.length
+					&& propertyValue === valueToAdd)) {
+				return;
+			}
+
+			propertyValueToSet = `${propertyValueToSet};${originalValueToAdd}`;
+		}
+
+		await this.setProjectPropertyValue(propertyName, propertyValueToSet);
+	}
+
+	/**
+	 * Removes value from project property where values are semicolon separated
+	 * @param propertyName Name of property
+	 * @param valueToRemove Value to remove from the project property
+	 * @param caseSensitive Determines if value to remove from collection should be compared case sensitively
+	 */
+	protected async removeValueFromCollectionProjectProperty(propertyName: string, valueToRemove: string, caseSensitive?: boolean): Promise<void> {
+		let propertyValueToSet = this.evaluateProjectPropertyValue(propertyName);
+		if (propertyValueToSet === undefined
+			|| propertyValueToSet.length < valueToRemove.length
+			|| valueToRemove.includes(';')) {
+			return;
+		}
+
+		const propertyValue = caseSensitive ? propertyValueToSet : propertyValueToSet.toLowerCase();
+		valueToRemove = caseSensitive ? valueToRemove : valueToRemove.toLowerCase();
+
+		if (propertyValue.length > valueToRemove.length) {
```

Comment:
Much of this logic (and that in "add") could be simplified (and therefore less prone to error) if we just convert the property value string to the object type it represents: an array.  Is this really worth it over some array operations that would be imperceptibly slower?  My gut says "no".

Reply:
Hm... you would need to run the whole parsing, which (given that implemented somewhat descent) would allocate an array (worth case allocate and copy log(n) arrays), then you would need to scan that array, which is _n_ operations, then remove an element from the array, which is another alloc and copy and finally process the array to format it as string, which is another _n-1_ operations, which internally is likely to allocate another character array (worst case log(n-1) arrays). All this will need to be then garbage collected. The actual string comparison is omitted from the equation as it will have to be done for this implementation regardless. Maybe I'm not familiar with the meaning of the word "imperceptibly"? I don't think we write code base on the "gut" feelings, there is a somewhat accurate complexity analysis that can be done, me thinks.

I would leave this as-is, but if owners of the code think that bunch of array allocations are okay (because this perhaps is not a performance-critical code path?), then we can change it. But saying that bunch of array allocations (including subsequent GC) are not that much slower than scanning a string 3 times does not sound as a legitimate argument to me.